machine/v1alpha1/types_openstack.go

```diff
@@ -0,0 +1,358 @@
+
+	// The machine ssh username
+	SshUserName string `json:"sshUserName,omitempty"`
+
+	// A networks object. Required parameter when there are multiple networks defined for the tenant.
+	// When you do not specify the networks parameter, the server attaches to the only network created for the current tenant.
+	Networks []NetworkParam `json:"networks,omitempty"`
+
+	// Create and assign additional ports to instances
+	Ports []PortOpts `json:"ports,omitempty"`
+
+	FloatingIP string `json:"floatingIP,omitempty"`
+
+	// The availability zone from which to launch the server.
+	AvailabilityZone string `json:"availabilityZone,omitempty"`
+
+	// The names of the security groups to assign to the instance
+	SecurityGroups []SecurityGroupParam `json:"securityGroups,omitempty"`
+
+	// The name of the secret containing the user data (startup script in most cases)
+	UserDataSecret *corev1.SecretReference `json:"userDataSecret,omitempty"`
+
+	// Whether the server instance is created on a trunk port or not.
+	Trunk bool `json:"trunk,omitempty"`
+
+	// Machine tags
+	// Requires Nova api 2.52 minimum!
+	Tags []string `json:"tags,omitempty"`
+
+	// Metadata mapping. Allows you to create a map of key value pairs to add to the server instance.
+	ServerMetadata map[string]string `json:"serverMetadata,omitempty"`
+
+	// Config Drive support
+	ConfigDrive *bool `json:"configDrive,omitempty"`
+
+	// The volume metadata to boot from
+	RootVolume *RootVolume `json:"rootVolume,omitempty"`
+
+	// The server group to assign the machine to.
+	ServerGroupID string `json:"serverGroupID,omitempty"`
+
+	// The server group to assign the machine to. A server group with that
+	// name will be created if it does not exist. If both ServerGroupID and
+	// ServerGroupName are non-empty, they must refer to the same OpenStack
+	// resource.
+	ServerGroupName string `json:"serverGroupName,omitempty"`
+
+	// The subnet that a set of machines will get ingress/egress traffic from
+	PrimarySubnet string `json:"primarySubnet,omitempty"`
+}
+
+type SecurityGroupParam struct {
+	// Security Group UID
+	UUID string `json:"uuid,omitempty"`
+	// Security Group name
+	Name string `json:"name,omitempty"`
+	// Filters used to query security groups in openstack
+	Filter SecurityGroupFilter `json:"filter,omitempty"`
+}
+
+type SecurityGroupFilter struct {
+	ID          string `json:"id,omitempty"`
+	Name        string `json:"name,omitempty"`
+	Description string `json:"description,omitempty"`
+	TenantID    string `json:"tenantId,omitempty"`
+	ProjectID   string `json:"projectId,omitempty"`
+	Limit       int    `json:"limit,omitempty"`
+	Marker      string `json:"marker,omitempty"`
+	SortKey     string `json:"sortKey,omitempty"`
+	SortDir     string `json:"sortDir,omitempty"`
+	Tags        string `json:"tags,omitempty"`
+	TagsAny     string `json:"tagsAny,omitempty"`
+	NotTags     string `json:"notTags,omitempty"`
+	NotTagsAny  string `json:"notTagsAny,omitempty"`
+}
```

Comment:
Would be nice to have at least some comments on each of these

Reply:
Most of these have literally no meaning in this context 🙄 In fact we no longer make any attempt to use them in MAPO. These and several other Filter objects like them were lifted wholesale from the OpenStack search API for their respective objects.